### PR TITLE
[WIP] KubeMQ (kubets) Plugin

### PR DESCRIPTION
**Pull request by Yomanz**
### What does this PR do?
It adds a plugin compatible with KubeMQ, specifically [kubets](https://github.com/Yomanz/Kubets)

### Motivation
I setup APM today and it was dope but I really needed a bit extra information from kubemq.

### Plugin Checklist

- [ ] Unit tests.
- [ ] TypeScript [definitions][1].
- [ ] TypeScript [tests][2].
- [ ] API [documentation][3].
- [ ] CircleCI [jobs/workflows][4].
- [x] Plugin is [exported][5].

[1]: https://github.com/DataDog/dd-trace-js/blob/master/index.d.ts
[2]: https://github.com/DataDog/dd-trace-js/blob/master/docs/test.ts
[3]: https://github.com/DataDog/dd-trace-js/blob/master/docs/API.md
[4]: https://github.com/DataDog/dd-trace-js/blob/master/.circleci/config.yml
[5]: https://github.com/DataDog/dd-trace-js/blob/master/packages/dd-trace/src/plugins/index.js

### Additional Notes
This would be SO much easier if it was in typescript :)
Issue: https://github.com/DataDog/dd-trace-js/issues/864

### CODE REVIEW

**Comment by rochdev** [comment]
Let me know when this is ready for review, but so far you are on the right track 👍

**Comment by Yomanz** [comment]
@rochdev A bit stuck.

I create the span for sending the actual request but I'm not 100% sure how to finish the span (when we receive a response for that message.
Would I need to pass something in the message so it can be extracted on the other end?

**Comment by rochdev** [comment]
I don't know enough about KubeMQ to answer specifically for it, so let me give a bit of background on how message queues are usually implemented.

For message queues with some kind of acknowledgement mechanism, in general there would be a callback or similar that can be used to finish the span. When there is no such callback, the only way to finish the span is right after the call to send the message. The same is true for the consumer. If it's just an event, then the span can be finished right before the handler is called. Otherwise if there is a callback it should be finished in the callback.

For context propagation, in general that would be done on the message. So you would create a span for the request in the producer, then attach that to messages somehow (attributes, headers, metadata, whatever is available). In the consumer, there would be 1 span per message since they are received individually. That span would also be activated on the scope of the handler and would be the parent of any other span created there since it's the root operation for that message.

A few examples: [amqp10](https://github.com/DataDog/dd-trace-js/blob/53d1f2131d61507aacd2fe1d7b157604c3d159ea/packages/datadog-plugin-amqp10/src/index.js), [amqplib](https://github.com/DataDog/dd-trace-js/blob/53d1f2131d61507aacd2fe1d7b157604c3d159ea/packages/datadog-plugin-amqplib/src/index.js), [pubsub](https://github.com/DataDog/dd-trace-js/blob/53d1f2131d61507aacd2fe1d7b157604c3d159ea/packages/datadog-plugin-google-cloud-pubsub/src/index.js), [rhea](https://github.com/DataDog/dd-trace-js/blob/53d1f2131d61507aacd2fe1d7b157604c3d159ea/packages/datadog-plugin-rhea/src/index.js)

**Comment by Yomanz** [comment]
Ah.

The pattern for the requests I'm attempting to implement is:

Send req -> Receive Request, Send data -> Sender receives the requested data.
If the receiver doesn't reply with data then an `INTERNAL 301` error is thrown, this means that the request timed-out and there was no reply.

I'll look into the `amqplib` plugin as I've used rabbitmq before :)

**Comment by rochdev** [comment]
Not sure I understand the above pattern. Does it mean the sender is waiting for the receiver to reply? If that's the case, usually this would be implemented using 2 messages, with each end acting as both a sender and a receiver, with a custom layer on top acting as a request/response handler. Is this the case or am I misunderstanding?

**Comment by Yomanz** [comment]
> 
> 
> Not sure I understand the above pattern. Does it mean the sender is waiting for the receiver to reply? If that's the case, usually this would be implemented using 2 messages, with each end acting as both a sender and a receiver, with a custom layer on top acting as a request/response handler. Is this the case or am I misunderstanding?

Sorry for the late reply.

Yes, there are multiple types of requests but the one I'm currently focusing on is as you described.

The sender waits for a reply from the receiver.

**Comment by rochdev** [comment]
Ok so in general the way it would work would be:

1) Service A produces a message and sends it to service B.
2) A span for the producer is started when the message is sent and finished when the message is acknowledged by either the broker or the consumer.
3) A span for the consumer is started when the message is received from either the broker or the producer and finished when acknowledged. That span would be a child of the producer span if the context was propagated by the producer.
4) Service B produces a message and sends it to service A.
5) A span for the producer is started when the message is sent and finished when the message is acknowledged by either the broker or the consumer.
6) A span for the consumer is started when the message is received from either the broker or the producer and finished when acknowledged. That span would be a child of the producer span if the context was propagated by the producer.

So basically, there would be 2 traces of 2 spans each. One for the request message, and one for the response message.

If you want everything to be in a single trace for the request, then the request/response pattern must be explicitly supported by either the broker or the library so that it can be instrumented automatically, otherwise it would need to be done manually.

What handles linking the request and response messages together?

**Comment by Yomanz** [comment]
When you request information it sends a GRPC request to KubeMQ, this sends a GRPC request to the receiver, you then send a response back to the sender.
The promise (on the request function from the sender) is then fulfilled with the data from the receiver.

**Comment by rochdev** [comment]
Does the above happen in a single request?

So basically, which one of these is true:

1) Producer sends a gRPC request and receives the data in the response.
2) Producer sends a gRPC request, and the consumer sends another request to the producer with the data.

If it's `1` I would definitely need more information as to how this work, as it would certainly sound like the behavior is not akin to a message queue.

It's also worth noting that we support gRPC out of the box so you should already be getting some level of instrumentation.

**Comment by Yomanz** [comment]
I'm horrific at explaining so enjoy this flowchart.
image
This is purely a Request<->Response flow at the moment as I plan to implement queues into kubets and this module prior to this initial flow.
I use the gRPC impelementation currently but it doesn't quite give enough information specifc to KubeMQ.
